src/main/BlockPool.cpp

```diff
@@ -0,0 +1,1216 @@
+    for (int i = 0; i < num_buckets; ++i) {
+      buckets[i].data = EMPTY_KEY;
+    }
+  }
+
+  /**
+   * Add an allocation (large object or medium object superblock) to the
+   * allocations hashtable.
+   */
+  void insert(uintptr_t pointer, size_t size) {
+    while (!try_insert(buckets, pointer, size)) {
+      // Remove partially inserted entries.
+      erase(pointer, size);
+      // Too many collisions, rebalance the hash table.
+      unsigned new_table_bits = num_bits + 1;
+      while (!rebalance(new_table_bits)) {
+        new_table_bits += 1;
+      }
+    }
+  }
+
+  /*
+   * Erases all hashtable entries for a pointer. The size determines
+   * how many hash buckets are removed.
+   */
+  void erase(uintptr_t pointer, size_t size) {
+    uintptr_t key = pointer >> LOW_BITS;
+    uintptr_t key_end = (pointer + (1 << size) + ((1 << LOW_BITS) - 1))
+        >> LOW_BITS;
+    bool first = true;
+    pointer &= ~uintptr_t{3};  // Mask out the flag bits for equality comparison.
+    do {
+      unsigned hash = pointer_hash(key << LOW_BITS);
+      for (int i = 0; i < MAX_COLLISIONS; ++i) {
+        HashBucket& bucket = buckets[hash];
+        if (bucket.DataPointer() == pointer) {
+          bucket.data = DELETED_KEY;
+          break;
+        }
+        if (bucket.data == EMPTY_KEY) {
+          // No more partial entries inserted, we are done.
+          return;
+        }
+        hash = probe_func(hash, i);
+      }
+      key += 1;
+    } while (key < key_end);
+  }
+
+  /*
+   * Find a bucket matching the pointer. Will find internal pointers.
+   */
+  HashBucket* search(uintptr_t candidate) {
+    unsigned hash = pointer_hash(candidate);
+    for (int i = 0; i < MAX_COLLISIONS; ++i) {
+      HashBucket* bucket = &buckets[hash];
+      if (bucket->data != EMPTY_KEY && bucket->data != DELETED_KEY) {
+        uintptr_t pointer = bucket->DataPointer();
+        if (bucket->data & 2) {
+          if (pointer <= candidate
+              && pointer + MEDIUM_SUPERBLOCK_SIZE > candidate) {
+            // Pointer is inside this superblock. Test if it is after the header.
+            if (candidate >= pointer + SUPERBLOCK_HEADER_SIZE) {
+              return bucket;
+            } else {
+              return nullptr;
+            }
+          }
+        } else if (pointer <= candidate
+              && (pointer + (1L << bucket->size)) > candidate) {
+          return bucket;
+        }
+      }
+      if (bucket->data == EMPTY_KEY) {
+        return nullptr;
+      }
+      lookup_collisions += 1;
+      hash = probe_func(hash, i);
+    }
+    return nullptr;
+  }
+
+  /**
+   * Attempts to rebalance the hashtable to use a new number of bits
+   * for hash keys. The rebalance fails if the number of collisions
+   * for a single entry exceeds MAX_REBALANCE_COLLISIONS.
+   * @return true if the rebalance succeeded.
+   */
+  bool rebalance(unsigned new_alloctable_bits) {
+    if (new_alloctable_bits > 29) {
```

Comment:
The hash key is usable up to 64 bits. This limit is a leftover from when my code was flaky and would eat up enormous amounts of memory and crash my machine fairly regularly. I put this limit in here so that the program would halt instead of eating all memory in the universe.

I will remove it!